src/GPS.cpp

```diff
@@ -0,0 +1,110 @@
+#include "GPS.hpp"
+#include <nmea0183/GPS.hpp>
+
+using namespace nmea0183;
+using namespace marnav;
+using namespace std;
+using namespace gps_base;
+
+GPS_SOLUTION_TYPES GPS::getPositionType(utils::optional<nmea::mode_indicator> mode)
+{
+    switch (*mode) {
+        case nmea::mode_indicator::invalid:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::autonomous:
+            return GPS_SOLUTION_TYPES::AUTONOMOUS;
+        case nmea::mode_indicator::differential:
+            return GPS_SOLUTION_TYPES::DIFFERENTIAL;
+        case nmea::mode_indicator::estimated:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::manual_input:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::simulated:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::data_not_valid:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::precise:
+            return GPS_SOLUTION_TYPES::AUTONOMOUS;
+        default:
+            return GPS_SOLUTION_TYPES::INVALID;
+    }
+}
+
+base::Time GPS::buildRockTime(utils::optional<nmea::time> const& optional_time,
+    utils::optional<nmea::date> const& optional_date)
+{
+    auto date = optional_date.value();
+    auto time = optional_time.value();
+    return base::Time::fromTimeValues(date.year() + 2000,
+        static_cast<int>(date.mon()),
+        date.day(),
+        time.hour(),
+        time.minutes(),
+        time.seconds(),
+        time.milliseconds(),
+        0);
+}
+
+Solution GPS::getSolution(nmea::rmc const& rmc, nmea::gsa const& gsa)
+{
+    Solution solution;
+    auto mode_indicator = rmc.get_mode_ind().value();
+    solution.positionType = getPositionType(mode_indicator);
+    if (solution.positionType != GPS_SOLUTION_TYPES::INVALID) {
```

Comment:
It seems that having a valid positionType and not having a latitude and/or longitude is possible, but in these cases we shouldnt report a valid gps solution. Maybe something like:

```
solution.time = buildRocktime(...);

auto position_type = getPositionType(mode_indicator);
auto latitude = rmc.get_latitude();
auto longitude = rmc.get_longitude();
if (position_type != GPS_SOLUTION_TYPES::INVALID && latitude.has_value() && longitude.has_value())
{

    solution.positionType = position_type;
    solution.latitude = latitude.value();
    solution.longitude = longitude.value();
    ... other stuff that are a part of the solution
}
else {
    solution.positionType = GPS_SOLUTION_TYPES::INVALID;
    ... add the nan lat long and other stuff that would describe an invalid solution
```